packages/integration_test/lib/integration_test.dart

```diff
@@ -11,61 +11,117 @@ import 'package:flutter_test/flutter_test.dart';
 import 'package:vm_service/vm_service_io.dart' as vm_io;
 
 import 'common.dart';
 import '_extension_io.dart' if (dart.library.html) '_extension_web.dart';
 import '_callback_io.dart' if (dart.library.html) '_callback_web.dart'
     as driver_actions;
+import 'src/constants.dart';
+import 'src/reporter.dart';
+
+bool _isUsingLegacyReporting = true;
+
+/// Executes a block that contains tests.
+///
+/// Example Usage:
+/// ```
+/// import 'package:flutter_test/flutter_test.dart';
+/// import 'package:integration_test/integration_test.dart';
+///
+/// void main() => run(_testMain);
+///
+/// void _testMain() {
+///   test('A test', () {
+///     expect(true, true);
+///   });
+/// }
+/// ```
+///
+/// The returned future will complete with the test results of the running
+/// [testMain]. These results will also be sent to native over the platform
+/// channel, unless [reportResultsToNative] is set to false.
+// TODO(jiahaog): Have stronger types for the returned success / failure result.
+Future<Map<String, Object>> run(FutureOr<void> Function() testMain,
+    {bool reportResultsToNative = true}) async {
+  assert(WidgetsBinding.instance == null);
+
+  _isUsingLegacyReporting = false;
+  final IntegrationTestWidgetsFlutterBinding binding =
+      IntegrationTestWidgetsFlutterBinding.ensureInitialized();
+
+  // Pipe detailed exceptions within [testWidgets] to `package:test`.
+  reportTestException = (FlutterErrorDetails details, String testDescription) {
+    registerException('Test $testDescription failed: $details');
+  };
+
+  final Completer<Map<String, Object>> resultsCompleter =
+      Completer<Map<String, Object>>();
+
+  await directRunTests(
+    testMain,
+    reporterFactory: (engine) => ResultReporter(engine, resultsCompleter),
+  );
+
+  final Map<String, Object> results = await resultsCompleter.future;
+
+  if (reportResultsToNative) {
+    await _reportResultsToNative(binding, results);
+  }
+  return results;
+}
 
-const String _success = 'success';
+String _formatFailureForPlatform(Failure failure) =>
+    '${failure.error} ${failure.details}';
+
+Future<void> _reportResultsToNative(
+    IntegrationTestWidgetsFlutterBinding binding,
+    Map<String, Object> results) async {
+  binding.results = results;
+  print('Test execution completed: ${binding.results}');
+
+  binding._allTestsPassed
+      .complete(!binding.results.values.any((val) => val is Failure));
+
+  try {
+    binding.callbackManager.cleanup();
+    await IntegrationTestWidgetsFlutterBinding._channel.invokeMethod<void>(
+      'allTestsFinished',
+      <String, dynamic>{
+        'results': {
+          for (final result in binding.results.entries)
+            result.key: result.value is Failure
+                ? _formatFailureForPlatform(result.value)
+                : result.value
+        }
+      },
+    );
+  } on MissingPluginException {
+    print('Warning: integration_test test plugin was not detected.');
```

Comment:
I think this should be treated as a more serious failure, and probably make us `exit` with some non-zero exit code.

The risk is that if this isn't set up right in the application, it will print what looks like a warning in a sea of log output and exit with zero when tests may have failed.

Reply:
We don't want the application to exit when it is invoked using flutter driver, as the native plugin is not used. The platform side plugin is only used when running with FTL or through Xcode. The responsibility of failing when the infrastructure is not set up properly (i.e. the native plugin is not installed) should probably be up to the host side invoker though, and probably not here.